src/lib.rs

```diff
@@ -2,23 +2,37 @@
 use pyo3::{prelude::*, wrap_pyfunction};
 
 use cpy_binder::export_cpy;
+use lazy_static::lazy_static;
+use std::sync::Mutex;
 
 struct NavigationManager {
     navigator: navigator_rs::Navigator,
 }
+
+lazy_static! {
+    static ref NAVIGATOR: Mutex<Option<NavigationManager>> = Mutex::new(None);
+}
+
 impl NavigationManager {
     // private constructor to prevent direct instantiation
-    fn new() -> NavigationManager {
-        NavigationManager {
-            navigator: navigator_rs::Navigator::new(),
+    fn new() -> &'static Mutex<Option<Self>> {
+        let mut st = NAVIGATOR.lock().unwrap();
+        if st.is_none() {
+            *st = Some(NavigationManager {
+                navigator: navigator_rs::Navigator::new(),
+            });
+            &NAVIGATOR
+        } else {
+            panic!("NavigationManager is already initialized")
         }
     }
 
-    fn get_instance() -> &'static mut NavigationManager {
-        static mut SINGLETON_INSTANCE: Option<NavigationManager> = None;
-
-        // use unsafe block to ensure thread safety
-        unsafe { SINGLETON_INSTANCE.get_or_insert_with(NavigationManager::new) }
+    fn get_instance() -> &'static Mutex<Option<Self>> {
+        if NAVIGATOR.lock().unwrap().is_some() {
+            &NAVIGATOR
+        } else {
+            panic!("NavigationManager must be initialized before use")
```

Comment:
we should init on the place of a panic.

Reply:
what about the new commit, get_instance can handle both conditions now.